### PR TITLE
Block reward rate

### PR DESCRIPTION
**Pull request by H-Pavel**
# Checklist:

- [ ] Code follows the style guidelines
- [ ] Documentation updated
- [ ] Corner cases evaluated and handled
- [ ] index.js exports updated
- [ ] Automatic tests added/updated
- [ ] Automatic tests pass
- [ ] Sanity check passed
- [ ] Self-review passed
- [ ] Version updated

### CODE REVIEW

File: src/constants.js [hunk 1]
```diff
@@ -34,3 +34,11 @@ export const FIRST_PHASE_DURATION = 2 * 60 * 60;
 export const ATLAS1_RELATIVE_STRENGTH = 1;
 export const ATLAS2_RELATIVE_STRENGTH = 4;
 export const ATLAS3_RELATIVE_STRENGTH = 12;
+
+/**
+ * Computed as:
+ * floor((Total AMB supply) * 2% / (blocks/year))
+ */
+export const DEFAULT_BLOCK_REWARD = '1146237435108134836';
```
[the remark this answers is not in the document]
Comment (H-Pavel) [reply src/constants.js]: Fixed

File: src/wrappers/block_rewards_wrapper.js [hunk 2]
```diff
@@ -22,4 +25,37 @@ export default class BlockRewardsWrapper extends GenesisContractWrapper {
   async getBaseReward() {
     return this.contract.methods.baseReward().call();
   }
+
+  convertRateToBaseReward(rate) {
+    if (rate < 0 || rate > 100) {
+      return -1;
```
[the remark this answers is not in the document]
Comment (H-Pavel) [reply src/wrappers/block_rewards_wrapper.js]: Fixed

File: src/wrappers/block_rewards_wrapper.js [hunk 3]
```diff
@@ -22,4 +25,37 @@ export default class BlockRewardsWrapper extends GenesisContractWrapper {
...
+    }
+
+    const totalSupply = new BN(TOTAL_AMB_SUPPLY);
+    const numberOfBlocksInYear = new BN(BLOCKS_PER_YEAR);
+    /* Percentage rate works with float numbers with up to 2 digits after the point */
+    const rewardPercent = totalSupply.div(numberOfBlocksInYear).divn(100);
```
[the remark this answers is not in the document]
Comment (H-Pavel) [reply src/wrappers/block_rewards_wrapper.js]: Fixed and simplified

File: src/wrappers/block_rewards_wrapper.js [hunk 4]
```diff
@@ -22,4 +25,37 @@ export default class BlockRewardsWrapper extends GenesisContractWrapper {
...
+    const percentageRate = rate * 100; /* Adjust percentage rate to contain digits after the point */
+
+    return rewardPercent.muln(percentageRate).divn(100);
```
[the remark this answers is not in the document]
Comment (H-Pavel) [reply src/wrappers/block_rewards_wrapper.js]: Fixed and simplified. Additinal div/mul by 100 was added for float numbers usage with two digits after the point through BN.js. After thinking I changed it to BigNumber.js usage, so all calculations are simplified now.

File: src/wrappers/block_rewards_wrapper.js [hunk 5]
```diff
@@ -22,4 +25,37 @@ export default class BlockRewardsWrapper extends GenesisContractWrapper {
...
+
+  convertBaseRewardToRate(baseReward) {
+    const totalSupply = new BN(TOTAL_AMB_SUPPLY);
+    const numberOfBlocksInYear = new BN(BLOCKS_PER_YEAR);
+    /* Percentage rate works with float numbers with up to 2 digits after the point */
+    const percentageRate = baseReward.mul(numberOfBlocksInYear).muln(100)
+      .muln(100)
+      .divRound(totalSupply);
+
+    const rate = percentageRate.toNumber() / 100; /* Adjust percentage rate to reveal digits after the point*/
```
[the remark this answers is not in the document]
Comment (H-Pavel) [reply src/wrappers/block_rewards_wrapper.js]: Same, as above - changed to BigNumber.js usage. No needed anymore.

File: src/wrappers/block_rewards_wrapper.js [hunk 6]
```diff
@@ -22,4 +25,37 @@ export default class BlockRewardsWrapper extends GenesisContractWrapper {
   async getBaseReward() {
     return this.contract.methods.baseReward().call();
   }
+
+  convertRateToBaseReward(rate) {
```
[the remark this answers is not in the document]
Comment (H-Pavel) [reply src/wrappers/block_rewards_wrapper.js]: I do not think, that we need new directory, as we have already used utils directory for additional functions. So I moved functionality to utils/block_reward_functions.js

File: test/wrappers/block_rewards_wrapper.js [hunk 7]
```diff
@@ -40,6 +44,36 @@ describe('Block Rewards Wrapper', () => {
     await expect(wrapper.getBaseReward()).to.eventually.equal(baseReward);
   });
 
+  it('convertRateToBaseReward converts percentage correctly', async () => {
```
[the remark this answers is not in the document]
Comment (H-Pavel) [reply test/wrappers/block_rewards_wrapper.js]: Changed to the wider test.

File: test/wrappers/block_rewards_wrapper.js [hunk 8]
```diff
@@ -40,6 +44,36 @@ describe('Block Rewards Wrapper', () => {
...
+
+    expect(wrapper.convertRateToBaseReward(rate).toString()).to.equal(reward);
+  });
+
+  it('convertRateToBaseReward rejects incorrect percentage', async () => {
+    expect(wrapper.convertRateToBaseReward(101)).to.equal(-1);
+    expect(wrapper.convertRateToBaseReward(-2)).to.equal(-1);
+  });
+
+  it('convertBaseRewardToRate returns correct percentage', async () => {
```
[the remark this answers is not in the document]
Comment (H-Pavel) [reply test/wrappers/block_rewards_wrapper.js]: Changed to the wider test.

File: test/wrappers/block_rewards_wrapper.js [hunk 9]
```diff
@@ -40,6 +44,36 @@ describe('Block Rewards Wrapper', () => {
...
+    let reword = new BN(DEFAULT_BLOCK_REWARD);
```
[the remark this answers is not in the document]
Comment (H-Pavel) [reply test/wrappers/block_rewards_wrapper.js]: Fixed

File: test/wrappers/block_rewards_wrapper.js [hunk 10]
```diff
@@ -40,6 +44,36 @@ describe('Block Rewards Wrapper', () => {
     await expect(wrapper.getBaseReward()).to.eventually.equal(baseReward);
   });
 
+  it('convertRateToBaseReward converts percentage correctly', async () => {
+    let rate = 2;
+    let reward = DEFAULT_BLOCK_REWARD;
+
+    expect(wrapper.convertRateToBaseReward(rate).toString()).to.equal(reward);
+
+    rate = 0.1;
```
[the remark this answers is not in the document]
Comment (H-Pavel) [reply test/wrappers/block_rewards_wrapper.js]: Done

File: src/wrappers/block_rewards_wrapper.js [hunk 11]
```diff
@@ -22,4 +25,37 @@ export default class BlockRewardsWrapper extends GenesisContractWrapper {
...
+    if (rate < 0 || rate > 100) {
+      return -1;
+    }
+
+    const totalSupply = new BN(TOTAL_AMB_SUPPLY);
+    const numberOfBlocksInYear = new BN(BLOCKS_PER_YEAR);
+    /* Percentage rate works with float numbers with up to 2 digits after the point */
```
[the remark this answers is not in the document]
Comment (H-Pavel) [reply src/wrappers/block_rewards_wrapper.js]: Fixed

File: test/utils/block_reward_functions.js [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+    expect(parseFloat(convertBaseRewardToRate(resultsToInputs[1.01]))).to.equal(1.01);
+    expect(parseFloat(convertBaseRewardToRate(resultsToInputs[1.3]))).to.equal(1.3);
+    expect(parseFloat(convertBaseRewardToRate(resultsToInputs[2]))).to.equal(2);
+    expect(parseFloat(convertBaseRewardToRate(resultsToInputs[2.5]))).to.equal(2.5);
+    expect(parseFloat(convertBaseRewardToRate(resultsToInputs[3]))).to.equal(3);
+    expect(parseFloat(convertBaseRewardToRate(resultsToInputs[5]))).to.equal(5);
+  });
+
+  it ('Roundtrip rate test', async () => {
+    const defaultRate = (Math.random() * 100).toFixed(2);
```
Comment (krzysztof-jelski) [inline test/utils/block_reward_functions.js]: Select a representative test case(s). Or run for the `resultsToInputs`. Or use property-based testing. But remove random. Consider that if the test failed the best we can do is take the value for which it failed, add a new test case and correct the bug.

File: test/utils/block_reward_functions.js [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+This Source Code Form is subject to the terms of the Mozilla Public License, v. 2.0. If a copy of the MPL was not distributed with this file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+This Source Code Form is “Incompatible With Secondary Licenses”, as defined by the Mozilla Public License, v. 2.0.
+*/
+
+import chai, {expect} from 'chai';
+import chaiAsPromised from 'chai-as-promised';
+import {convertRateToBaseReward, convertBaseRewardToRate} from '../../src/utils/block_reward_functions';
+
+chai.use(chaiAsPromised);
```
Comment (krzysztof-jelski) [inline test/utils/block_reward_functions.js]: Not needed anymore.

File: test/utils/block_reward_functions.js [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+chai.use(require('sinon-chai'));
```
Comment (krzysztof-jelski) [inline test/utils/block_reward_functions.js]: Not needed anymore.

File: test/utils/block_reward_functions.js (hunk 12 above)
Comment (H-Pavel) [reply test/utils/block_reward_functions.js]: Fixed

File: test/utils/block_reward_functions.js (hunk 13 above)
Comment (H-Pavel) [reply test/utils/block_reward_functions.js]: Fixed

File: test/utils/block_reward_functions.js (hunk 14 above)
Comment (H-Pavel) [reply test/utils/block_reward_functions.js]: Fixed

**Pull request merged**